File: apps/ios/SitrCore/Sources/SitrCore/SyncClient.swift
/// Sync client — pull → merge → push against the one documented endpoint
/// (docs/data-flow.md, docs/sync-protocol.md).
/// Port of extension/src/lib/sync/client.ts.
///
/// THIS FILE'S TRANSPORT IS THE iOS APP'S ONLY NETWORK CALL SITE —
/// data-flow.md invites auditors to verify that claim; keep it true.
///
/// INVARIANT: sync outcomes touch ONLY the sync status and household
/// state. They never touch protection status — filtering is local, so a
/// dead server leaves protection fully intact.
///
/// Rollback detection: we remember the highest household rev ever
/// decrypted (`maxSeenRev`). An authenticated blob with a lower rev is an
/// error ("server returned an older state"), surfaced but never applied.
import Foundation

public struct SyncHTTPResponse {
    public let status: Int
    public let etagHeader: String?
    public let body: Data

    public init(status: Int, etagHeader: String?, body: Data) {
        self.status = status
        self.etagHeader = etagHeader
        self.body = body
    }
}

/// Injected transport — the app wires URLSession in exactly one place;
/// tests inject a stub. Throwing means "offline".
public protocol SyncTransport {
    func request(
        method: String,
        url: URL,
        headers: [String: String],
        body: Data?
    ) async throws -> SyncHTTPResponse
}

public struct SyncInput {
    public let rootSecret: Data
    /// The locally applied state (may be ahead of the server's).
    public let local: HouseholdState
    /// Highest rev this device has ever decrypted from the server.
    public let maxSeenRev: Int
    public let deviceId: String
    /// Signed subscription token; sent only when creating a household.
    /// The iOS app is join-only and never sets this — the parameter exists
    /// for protocol parity (docs/sync-protocol.md §Entitlement).
    public let entitlement: String?

    public init(
        rootSecret: Data,
        local: HouseholdState,
        maxSeenRev: Int,
        deviceId: String,
        entitlement: String? = nil
    ) {
        self.rootSecret = rootSecret
        self.local = local
        self.maxSeenRev = maxSeenRev
        self.deviceId = deviceId
        self.entitlement = entitlement
    }
}

public struct SyncOutcome: Equatable {
    /// The state the caller must apply + persist (merge result).
    public let state: HouseholdState
    public let maxSeenRev: Int
    public let status: SyncStatus
    /// ETag of the server copy after this sync (diagnostics only).
    public let etag: Int?
}

public struct SyncClient {
    public static let defaultBaseURL = URL(string: "https://sync.sitrshield.com")!

    let transport: SyncTransport
    let now: () -> Double
    let baseURL: URL

    public init(
        transport: SyncTransport,
        now: @escaping () -> Double,
        baseURL: URL = SyncClient.defaultBaseURL
    ) {
        self.transport = transport
        self.now = now
        self.baseURL = baseURL
    }

    static func parseEtag(_ raw: String?) -> Int? {
        guard let raw else { return nil }
        let trimmed = raw.trimmingCharacters(in: .whitespaces)
        guard
            trimmed.count >= 3, trimmed.hasPrefix("\""), trimmed.hasSuffix("\""),
            case let digits = String(trimmed.dropFirst().dropLast()),
            digits.count <= 15, !digits.isEmpty,
            digits.allSatisfy({ $0.isASCII && $0.isNumber })
        else { return nil }
        return Int(digits)
    }

    struct Remote {
        let state: HouseholdState?
        let etag: Int?
    }

    enum PullError: Error {
        case offline
        case message(String)
    }

    func blobURL(_ keys: SyncCrypto.HouseholdKeys) -> URL {
        baseURL.appendingPathComponent("v1/blob/\(keys.householdId)")
    }

    func pull(keys: SyncCrypto.HouseholdKeys) async -> Result<Remote, PullError> {
        let response: SyncHTTPResponse
        do {
            response = try await transport.request(
                method: "GET",
                url: blobURL(keys),
                headers: ["Authorization": "Bearer \(keys.authToken)"],
                body: nil
            )
        } catch {
            return .failure(.offline)
        }
        if response.status == 404 { return .success(Remote(state: nil, etag: nil)) }
        guard (200..<300).contains(response.status) else {
            return .failure(.message("server responded \(response.status)"))
        }
        let etag = Self.parseEtag(response.etagHeader)
        switch Household.openState(blob: response.body, encKey: keys.encKey) {
        case .failure(let e): return .failure(.message(e.message))
        case .success(let state): return .success(Remote(state: state, etag: etag))
        }
    }

    enum PushError: Error {
        case offline
        case conflict
        case message(String)
    }

    func push(
        keys: SyncCrypto.HouseholdKeys,
        state: HouseholdState,
        etag: Int?,
        entitlement: String?
    ) async -> Result<Int, PushError> {
        let sealed = Household.sealState(state, encKey: keys.encKey)
        guard case .success(let blob) = sealed else {
            if case .failure(let e) = sealed { return .failure(.message(e.message)) }
            return .failure(.message("seal failed"))
        }
        var headers = ["Authorization": "Bearer \(keys.authToken)"]
        if let etag {
            headers["If-Match"] = "\"\(etag)\""
        } else {
            headers["If-None-Match"] = "*"
        }
        if let entitlement, !entitlement.isEmpty {
            headers["X-Sitr-Entitlement"] = entitlement
        }
        let response: SyncHTTPResponse
        do {
            response = try await transport.request(
                method: "PUT",
                url: blobURL(keys),
                headers: headers,
                body: blob
            )
        } catch {
            return .failure(.offline)
        }
        if response.status == 409 { return .failure(.conflict) }
        guard (200..<300).contains(response.status) else {
            return .failure(.message("server responded \(response.status)"))
        }
        return .success(Self.parseEtag(response.etagHeader) ?? 0)
    }

    /// One full sync round. Never throws. The returned state is always safe
    /// to apply: merged, sanitized (via openState), and rollback-checked.
    public func syncOnce(_ input: SyncInput) async -> SyncOutcome {
        func failed(_ error: String, offline: Bool = false) -> SyncOutcome {
            SyncOutcome(
                state: input.local,
                maxSeenRev: input.maxSeenRev,
                status: SyncStatus(state: offline ? .offline : .error, error: error),
                etag: nil
            )
        }

        let derived = SyncCrypto.deriveKeys(rootSecret: input.rootSecret)
        guard case .success(let keys) = derived else {
            if case .failure(let e) = derived { return failed(e.message) }
            return failed("key derivation failed")
        }

        enum AttemptError: Error {
            case retry
            case offline
            case message(String)
        }

        func attempt() async -> Result<SyncOutcome, AttemptError> {
            let remote: Remote
            switch await pull(keys: keys) {
            case .failure(.offline): return .failure(.offline)
            case .failure(.message(let m)): return .failure(.message(m))
            case .success(let r): remote = r
            }

            var merged = input.local
            var maxSeen = input.maxSeenRev
            if let remoteState = remote.state {
                if remoteState.rev < input.maxSeenRev {
                    return .failure(
                        .message(
                            "server returned an older household state than previously seen — refusing to apply it"
                        ))
                }
                maxSeen = max(maxSeen, remoteState.rev)
                merged = Household.merge(input.local, remoteState)
            }
            // Push only when the server copy differs from the merge result.
            if let remoteState = remote.state, remoteState.rev == merged.rev,
                remoteState == merged
            {
                return .success(
                    SyncOutcome(
                        state: merged,
                        maxSeenRev: max(maxSeen, merged.rev),
                        status: SyncStatus(state: .ok, lastSuccessAt: now()),
                        etag: remote.etag
                    ))
            }
            let toPush: HouseholdState
            if let remoteState = remote.state, remoteState.rev >= merged.rev,
                remoteState != merged
            {
                toPush = Household.bumpRev(merged, deviceId: input.deviceId, now: now())
            } else {
                toPush = merged
            }
            switch await push(
                keys: keys, state: toPush, etag: remote.etag,
                entitlement: input.entitlement)
            {
            case .failure(.conflict): return .failure(.retry)
            case .failure(.offline): return .failure(.offline)
            case .failure(.message(let m)): return .failure(.message(m))
            case .success(let etag):
                return .success(
                    SyncOutcome(
                        state: toPush,
                        maxSeenRev: max(maxSeen, toPush.rev),
                        status: SyncStatus(state: .ok, lastSuccessAt: now()),
                        etag: etag
                    ))
            }
        }

        switch await attempt() {
        case .success(let outcome): return outcome
        case .failure(.offline): return failed("offline", offline: true)
        case .failure(.message(let m)): return failed(m)
        case .failure(.retry):
            // One concurrent-write retry: re-pull, re-merge, re-push.
            switch await attempt() {
            case .success(let outcome): return outcome
            case .failure(.retry): return failed("repeated version conflicts")
            case .failure(.offline): return failed("offline", offline: true)
            case .failure(.message(let m)): return failed(m)
            }
        }
    }
}

File: apps/android/core/src/main/kotlin/com/sitrshield/core/sync/SyncClient.kt
package com.sitrshield.core.sync

import com.sitrshield.core.SitrResult
import com.sitrshield.core.household.Household
import com.sitrshield.core.household.HouseholdState
import java.io.IOException

/**
 * Sync client — pull → merge → push against the one documented endpoint
 * (docs/data-flow.md, docs/sync-protocol.md).
 * Port of extension/src/lib/sync/client.ts.
 *
 * THIS CLIENT'S TRANSPORT IS THE ANDROID APP'S ONLY HTTP CALL SITE —
 * the app implements SyncTransport in exactly one place (SyncHttp);
 * data-flow.md invites auditors to verify that claim; keep it true.
 *
 * INVARIANT: sync outcomes touch ONLY the sync status and household
 * state. They never touch protection status — filtering is local, so a
 * dead server leaves protection fully intact.
 *
 * Rollback detection: we remember the highest household rev ever
 * decrypted (`maxSeenRev`). An authenticated blob with a lower rev is an
 * error, surfaced but never applied.
 */
class SyncHttpResponse(
    val status: Int,
    val etagHeader: String?,
    val body: ByteArray,
)

/** Throw IOException to signal "offline"; anything else is a server error. */
interface SyncTransport {
    @Throws(IOException::class)
    fun request(
        method: String,
        url: String,
        headers: Map<String, String>,
        body: ByteArray?,
    ): SyncHttpResponse
}

data class SyncInput(
    val rootSecret: ByteArray,
    /** The locally applied state (may be ahead of the server's). */
    val local: HouseholdState,
    /** Highest rev this device has ever decrypted from the server. */
    val maxSeenRev: Int,
    val deviceId: String,
    /** Signed subscription token; sent only on household creation. */
    val entitlement: String? = null,
)

data class SyncOutcome(
    /** The state the caller must apply + persist (merge result). */
    val state: HouseholdState,
    val maxSeenRev: Int,
    val status: SyncStatus,
    /** ETag of the server copy after this sync (diagnostics only). */
    val etag: Int?,
)

class SyncClient(
    private val transport: SyncTransport,
    private val now: () -> Double,
    private val baseUrl: String = DEFAULT_BASE_URL,
) {
    companion object {
        const val DEFAULT_BASE_URL = "https://sync.sitrshield.com"

        internal fun parseEtag(raw: String?): Int? {
            val trimmed = raw?.trim() ?: return null
            val match = Regex("^\"(\\d{1,15})\"$").find(trimmed) ?: return null
            return match.groupValues[1].toIntOrNull()
        }
    }

    private class Remote(val state: HouseholdState?, val etag: Int?)

    private sealed class AttemptError {
        object Retry : AttemptError()
        object Offline : AttemptError()
        class Message(val message: String) : AttemptError()
    }

    private fun blobUrl(keys: SyncCrypto.HouseholdKeys): String =
        "$baseUrl/v1/blob/${keys.householdId}"

    private fun pull(keys: SyncCrypto.HouseholdKeys): SitrResult<Remote> {
        val response = try {
            transport.request(
                "GET", blobUrl(keys),
                mapOf("Authorization" to "Bearer ${keys.authToken}"),
                null,
            )
        } catch (_: IOException) {
            return SitrResult.Err(OFFLINE)
        }
        if (response.status == 404) return SitrResult.Ok(Remote(null, null))
        if (response.status !in 200..299) {
            return SitrResult.Err("server responded ${response.status}")
        }
        val etag = parseEtag(response.etagHeader)
        return when (val opened = Household.openState(response.body, keys.encKey)) {
            is SitrResult.Err -> opened
            is SitrResult.Ok -> SitrResult.Ok(Remote(opened.value, etag))
        }
    }

    private fun push(
        keys: SyncCrypto.HouseholdKeys,
        state: HouseholdState,
        etag: Int?,
        entitlement: String?,
    ): SitrResult<Int> {
        val sealed = when (val r = Household.sealState(state, keys.encKey)) {
            is SitrResult.Err -> return r
            is SitrResult.Ok -> r.value
        }
        val headers = buildMap {
            put("Authorization", "Bearer ${keys.authToken}")
            if (etag != null) put("If-Match", "\"$etag\"") else put("If-None-Match", "*")
            if (!entitlement.isNullOrEmpty()) put("X-Sitr-Entitlement", entitlement)
        }
        val response = try {
            transport.request("PUT", blobUrl(keys), headers, sealed)
        } catch (_: IOException) {
            return SitrResult.Err(OFFLINE)
        }
        if (response.status == 409) return SitrResult.Err(CONFLICT)
        if (response.status !in 200..299) {
            return SitrResult.Err("server responded ${response.status}")
        }
        return SitrResult.Ok(parseEtag(response.etagHeader) ?: 0)
    }

    /**
     * One full sync round. Never throws. The returned state is always safe
     * to apply: merged, sanitized (via openState), and rollback-checked.
     */
    fun syncOnce(input: SyncInput): SyncOutcome {
        fun failed(error: String, offline: Boolean = false): SyncOutcome =
            SyncOutcome(
                state = input.local,
                maxSeenRev = input.maxSeenRev,
                status = SyncStatus(
                    state = if (offline) SyncStatus.State.OFFLINE else SyncStatus.State.ERROR,
                    error = error,
                ),
                etag = null,
            )

        val keys = when (val derived = SyncCrypto.deriveKeys(input.rootSecret)) {
            is SitrResult.Err -> return failed(derived.message)
            is SitrResult.Ok -> derived.value
        }

        fun attempt(): Any /* SyncOutcome | AttemptError */ {
            val remote = when (val pulled = pull(keys)) {
                is SitrResult.Err ->
                    return if (pulled.message == OFFLINE) AttemptError.Offline
                    else AttemptError.Message(pulled.message)
                is SitrResult.Ok -> pulled.value
            }

            var merged = input.local
            var maxSeen = input.maxSeenRev
            val remoteState = remote.state
            if (remoteState != null) {
                if (remoteState.rev < input.maxSeenRev) {
                    return AttemptError.Message(
                        "server returned an older household state than previously seen — refusing to apply it"
                    )
                }
                maxSeen = maxOf(maxSeen, remoteState.rev)
                merged = Household.merge(input.local, remoteState)
            }
            // Push only when the server copy differs from the merge result.
            if (remoteState != null && remoteState.rev == merged.rev && remoteState == merged) {
                return SyncOutcome(
                    state = merged,
                    maxSeenRev = maxOf(maxSeen, merged.rev),
                    status = SyncStatus(SyncStatus.State.OK, lastSuccessAt = now()),
                    etag = remote.etag,
                )
            }
            val toPush =
                if (remoteState != null && remoteState.rev >= merged.rev && remoteState != merged)
                    Household.bumpRev(merged, input.deviceId, now())
                else merged
            return when (val pushed = push(keys, toPush, remote.etag, input.entitlement)) {
                is SitrResult.Err -> when (pushed.message) {
                    CONFLICT -> AttemptError.Retry
                    OFFLINE -> AttemptError.Offline
                    else -> AttemptError.Message(pushed.message)
                }
                is SitrResult.Ok -> SyncOutcome(
                    state = toPush,
                    maxSeenRev = maxOf(maxSeen, toPush.rev),
                    status = SyncStatus(SyncStatus.State.OK, lastSuccessAt = now()),
                    etag = pushed.value,
                )
            }
        }

        return when (val first = attempt()) {
            is SyncOutcome -> first
            AttemptError.Offline -> failed("offline", offline = true)
            is AttemptError.Message -> failed(first.message)
            AttemptError.Retry ->
                // One concurrent-write retry: re-pull, re-merge, re-push.
                when (val second = attempt()) {
                    is SyncOutcome -> second
                    AttemptError.Retry -> failed("repeated version conflicts")
                    AttemptError.Offline -> failed("offline", offline = true)
                    is AttemptError.Message -> failed(second.message)
                    else -> failed("unexpected sync state")
                }
            else -> failed("unexpected sync state")
        }
    }
}

private const val OFFLINE = "offline"
private const val CONFLICT = "conflict"
